### ISSUE TITLE
Handle materials in editor

### ISSUE THREAD

**Issue by freezy** [freezy/VisualPinball.Engine#79]
It's a bit unclear to me how we'll handle materials, so the main goal of this issue is to open the discussion.

![image](https://user-images.githubusercontent.com/70426/84543063-eb6fde00-acfa-11ea-855f-b5de4e21d443.png)

A material in VPX basically defines which shader to use and how to use it. Textures aren't part of the material, they are assigned to the object individually, however, there are physics properties as well.

Now, in order to fully benefit from Unity's renderer, we need to use Unity's shaders, which are obviously a very different beast.

So the first question would be: If authors use Unity just for authoring with the goal of using the VPX player, do we provide a way to deal with materials? There is no way of providing an accurate visualization, so what we could do is the following:

- When a Unity material is created or edited, create or update a VPX material as well. There are a bunch of parameters such as colors or shininess than can be converted in order achieve a similar look.
- Provide a VPX materials editor clone where an author can fine-tune the VPX materials. Knowing those won't affect Unity's shaders. 
- The VPX materials should be "lockable", i.e. they then won't be updated when corresponding Unity material updates.

In short, authors should be able to apply Unity materials to their meshes without any restrictions, but since we still want to be able to play the produced `.vpx` file in VPX, they should be converted and editable.

How we internally handle all that, which of Unity's render pipelines we support, how we write the Unity materials to the `.vpx` file etc will be part of another issue.

Thoughts?

**Comment by rbxnk** [comment]
My vote would be to do the following:

* Get as close as we can to VPX's look with a set of materials we create in Unity, drive those materials properties solely from the vpx material data.
* Introduce some new VPE only fields for these materials that allow authors to define renderer-specific material properties. Authors might want to tweak the look of something if its pbr vs a more traditional shader, or raytraced. But all this still using materials we provide that takes their values from the VPE item fields.

This would provide enough control to get us rolling, without having to solve if/how we pack in unity assets yet.

**Comment by freezy** [comment]
So the hard requirements from my side would be the following:

- When importing a vanilla `.vpx` file, materials should look somewhat similar in VPE compared to VPX's renderer.
- When creating a table in VPE, materials should also look similar in VPX.
- In VPX, authors should be able to tweak the material without changing the VPE material (given it's already "tweaked").
- In VPE, authors should have full access to all the shaders (incl shadergraph and custom shaders) when working with materials (how we convert those to VPX materials is another question).
- When loading a `.vpx` file into Unity, the used materials should be editable

The last point is non-trivial, because shaders need to be compiled, which we can't do runtime. So I think what we'll need to do is include the source assets as well as the compiled package in the `.vpx` file. 

My first instinct would be to try to hook into Unity's build system and just grab everything we don't already store in the `.vpx` file and put it into a separate path of the `.vpx` (it's like a virtual file system). And for the source assets we would do the same when importing.

Lastly, a few people have asked to provide an unpacked file structure for working with VPX table, so they could be hosted as projects on GitHub where people can submit PRs for mods and whatnot. So having separate files in either Unity's asset format or another one (to complicate things even more, but be more agnostic) is a strong point as well.

As mentioned already, writing the importer was my first experience with Unity, and I wouldn't mind getting it replaced altogether. We can start simple and iterate later.

**Comment by freezy** [comment]
@rbxnk Just re-read this and noticed that I didn't respond to your suggestion. What you've described is fine, we can iterate from there.

**Comment by freezy** [comment]
Implemented in #122.
